src/libtiled/mapwriter.cpp

```diff
@@ -648,41 +649,105 @@ void MapWriterPrivate::writeTileLayer(QXmlStreamWriter &w,
         QRect bounds = tileLayer.bounds().translated(-tileLayer.position());
         startX = bounds.left();
         startY = bounds.top();
-        endX = bounds.right();
-        endY = bounds.bottom();
-    }
+        endX = bounds.right() + 1;
+        endY = bounds.bottom() + 1;
+
+        int chunkStartX = startX;
+        int chunkStartY = startY;
+        int resetX = chunkStartX;
+
+        startX = (startX < 0) ? (startX + 1) / CHUNK_SIZE - 1 : startX / CHUNK_SIZE;
+        startY = (startY < 0) ? (startY + 1) / CHUNK_SIZE - 1 : startY / CHUNK_SIZE;
+        endX = (endX < 0) ? (endX + 1) / CHUNK_SIZE - 1 : endX / CHUNK_SIZE;
+        endY = (endY < 0) ? (endY + 1) / CHUNK_SIZE - 1 : endY / CHUNK_SIZE;
+
+        for (int y = startY; y < endY; ++y) {
+            for (int x = startX; x < endX; ++x) {
+                if (tileLayer.findChunk(chunkStartX, chunkStartY)) {
+                    w.writeStartElement(QLatin1String("chunk"));
+                    w.writeAttribute(QLatin1String("startx"), QString::number(chunkStartX));
+                    w.writeAttribute(QLatin1String("starty"), QString::number(chunkStartY));
+
+                    if (mLayerDataFormat == Map::XML) {
+                        for (int y1 = 0; y1 < CHUNK_SIZE; y1++) {
```

Comment:
Maybe use `cellY` here (or `chunkY` before)? I find adding numbers to variables to prevent name clashes a bit confusing often.

Reply:
Yes, but actually this whole code should not be duplicated in the first place. It should be moved into a function taking the QXmlStreamWriter, the tile layer and a `QRect` specifying which part of the layer to store. That function can then be called either once for the whole tile layer, or repeatedly for each chunk, depending on whether the map is infinite.